clients/src/main/java/org/apache/kafka/common/metrics/IntGaugeSuite.java

```diff
@@ -0,0 +1,140 @@
+                         int maxEntries) {
+        this.log = log;
+        this.suiteName = suiteName;
+        this.metrics = metrics;
+        this.metricNameCalculator = metricNameCalculator;
+        this.maxEntries = maxEntries;
+        this.map = new HashMap<>(1);
+        this.closed = false;
+        log.trace("{}: created new gauge suite with maxEntries = {}.",
+            suiteName, maxEntries);
+    }
+
+    public synchronized void increment(K shortName) {
+        if (closed) {
+            log.warn("{}: Attempted to increment {}, but the GaugeSuite was closed.",
+                suiteName, shortName.toString());
+            return;
+        }
+        StoredIntGauge gauge = map.get(shortName);
+        if (gauge == null) {
+            if (map.size() == maxEntries) {
+                log.warn("{}: Attempted to increment {}, but there are already {} entries.",
+                    suiteName, shortName.toString(), maxEntries);
+                return;
+            }
+            MetricName metricName = metricNameCalculator.apply(shortName);
+            gauge = new StoredIntGauge(metricName);
+            metrics.addMetric(metricName, gauge);
+            log.trace("{}: Added a new metric {}", suiteName, shortName.toString());
+            map.put(shortName, gauge);
+        }
+        gauge.value++;
+    }
+
+    public synchronized void decrement(K shortName) {
+        if (closed) {
+            log.warn("{}: Attempted to decrement {}, but the gauge suite was closed.",
+                suiteName, shortName.toString());
+            return;
+        }
+        StoredIntGauge gauge = map.get(shortName);
+        if (gauge == null) {
+            log.warn("{}: Attempted to decrement {}, but no such metric was registered.",
+                suiteName, shortName.toString());
+        } else {
+            gauge.value--;
+            if (gauge.value == 0) {
+                log.trace("{}: Removing {}.", suiteName, shortName.toString());
+                metrics.removeMetric(gauge.metricName);
+                map.remove(shortName);
+            } else {
+                log.trace("{}: Removed a reference to {}.  {} reference(s) remaining.",
+                    suiteName, shortName.toString(), gauge.value);
+            }
+        }
+    }
+
+    @Override
+    public synchronized void close() {
+        if (!closed) {
+            closed = true;
+            int prevSize = map.values().size();
+            for (StoredIntGauge gauge : map.values()) {
+                metrics.removeMetric(gauge.metricName);
+            }
+            log.trace("{}: closed {} metric(s).", suiteName, prevSize);
+            map.clear();
+        } else {
+            log.trace("{}: gauge suite is already closed.", suiteName);
+        }
+    }
+
+    public int maxEntries() {
+        return maxEntries;
+    }
+
+    // visible for testing only
+    synchronized void visit(Consumer<Map<K, StoredIntGauge>> visitor) {
```

Comment:
How is this better than a method that returns the map? Is it because we pass the map with a lock held? Could we expose a map get method instead? Calling callbacks with locks held tends to cause problems and I'm a bit concerned that someone will see this method and use it outside of tests.

Reply:
I fixed it to just return a new map for simplicity.